pkg/tempo/automaticloggingprocessor/automaticloggingprocessor.go

```diff
@@ -0,0 +1,242 @@
+	cfg          *AutomaticLoggingConfig
+	lokiChan     chan<- api.Entry
+	done         atomic.Bool
+
+	logger log.Logger
+}
+
+func newTraceProcessor(nextConsumer consumer.TracesConsumer, cfg *AutomaticLoggingConfig) (component.TracesProcessor, error) {
+	logger := log.With(util.Logger, "component", "tempo automatic logging")
+
+	if nextConsumer == nil {
+		return nil, componenterror.ErrNilNextConsumer
+	}
+
+	if !cfg.Roots && !cfg.Processes && !cfg.Spans {
+		return nil, errors.New("automaticLoggingProcessor requires one of roots, processes, or spans to be enabled")
+	}
+
+	cfg.Overrides.LokiTag = override(cfg.Overrides.LokiTag, defaultLokiTag)
+	cfg.Overrides.ServiceKey = override(cfg.Overrides.ServiceKey, defaultServiceKey)
+	cfg.Overrides.SpanNameKey = override(cfg.Overrides.SpanNameKey, defaultSpanNameKey)
+	cfg.Overrides.StatusKey = override(cfg.Overrides.StatusKey, defaultStatusKey)
+	cfg.Overrides.DurationKey = override(cfg.Overrides.DurationKey, defaultDurationKey)
+	cfg.Overrides.TraceIDKey = override(cfg.Overrides.TraceIDKey, defaultTraceIDKey)
+
+	return &automaticLoggingProcessor{
+		nextConsumer: nextConsumer,
+		cfg:          cfg,
+		logger:       logger,
+		done:         atomic.Bool{},
+	}, nil
+}
+
+func (p *automaticLoggingProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	rsLen := td.ResourceSpans().Len()
+	for i := 0; i < rsLen; i++ {
+		rs := td.ResourceSpans().At(i)
+		ilsLen := rs.InstrumentationLibrarySpans().Len()
+
+		var svc string
+		svcAtt, ok := rs.Resource().Attributes().Get(conventions.AttributeServiceName)
+		if ok {
+			svc = svcAtt.StringVal()
+		}
+
+		for j := 0; j < ilsLen; j++ {
+			ils := rs.InstrumentationLibrarySpans().At(j)
+			spanLen := ils.Spans().Len()
+
+			lastTraceID := ""
+			for k := 0; k < spanLen; k++ {
+				span := ils.Spans().At(k)
+				traceID := span.TraceID().HexString()
+
+				if p.cfg.Spans {
+					p.exportToLoki(typeSpan, traceID, p.spanKeyVals(span, svc)...)
+				}
+
+				if p.cfg.Roots && span.ParentSpanID().IsEmpty() {
+					p.exportToLoki(typeRoot, traceID, p.spanKeyVals(span, svc)...)
+				}
+
+				if p.cfg.Processes && lastTraceID != traceID {
+					lastTraceID = traceID
+					p.exportToLoki(typeProcess, traceID, p.processKeyVals(rs.Resource(), svc)...)
+				}
+			}
+		}
+	}
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *automaticLoggingProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{}
+}
+
+// Start is invoked during service startup.
+func (p *automaticLoggingProcessor) Start(ctx context.Context, _ component.Host) error {
+	loki := ctx.Value(contextkeys.Loki).(*loki.Loki)
+	if loki == nil {
+		return fmt.Errorf("key %s does not contain a Loki instance", contextkeys.Loki)
+	}
+	lokiInstance := loki.Instance(p.cfg.LokiName)
+	if lokiInstance == nil {
+		return fmt.Errorf("loki instance %s not found", p.cfg.LokiName)
+	}
+	p.lokiChan = lokiInstance.Promtail().Client().Chan()
```

Comment:
This isn't going to work if the promtail restarts 😞 We'll need this to use a channel on the instance (which should be immutable) and have the Loki instance forward from the active promtail over the instance's channel.

Reply:
Yeah, I did some mild testing and it seemed to work, but my expectation is that in the general case it would not. I was considering:

- Shutdown Tempo subsystem
- Reload Loki 
- Reload Tempo

I think this would work.

Reply:
It'd work, but I'd prefer to try to get this to work with the existing ApplyConfig mechanism and not introducing shutdowns if we can. What were your concerns with forwarding the channel?

Reply:
@rfratto have a first pass attempt at doing this with the `ApplyConfig`. Let me know what you think.